File: 백준/Silver/21736. 헌내기는 친구가 필요해/헌내기는 친구가 필요해.cc
#include <iostream>
#include <vector>
#include <array>
using namespace std;

array<int, 4> dx = {0, 0, -1, 1};
array<int, 4> dy = {1, -1, 0, 0};
const int V = 4;
int count = 0;

bool is_valid(int x, int y);
void dfs(int x, int y);

int n, m;
vector<string> field;
vector<vector<bool>> visited;
int main() {
    cin >> n >> m;
    field.assign(n, "");
    visited.assign(n, vector<bool>(m, false));

    int si = 0, sj = 0;
    for (int i = 0; i < n; i++) {
        cin >> field[i];
        for (int j = 0; j < m; j++) {
            if (field[i][j] == 'I') {
                si = i;
                sj = j;
            }
        }
    }

    dfs(si, sj);
    if (::count > 0) cout << ::count;
    else cout << "TT";
}

bool is_valid(const int x, const int y) {
    if (x < 0 || x >= n || y < 0 || y >= m || visited[x][y])
        return false;
    return true;
}

void dfs(int x, int y) {
    visited[x][y] = true;
    if (field[x][y] == 'P') ::count += 1;

    for (int v = 0; v < V; v++) {
        int next_x = x + dx[v];
        int next_y = y + dy[v];

        if (is_valid(next_x, next_y) && field[next_x][next_y] != 'X') {
            dfs(next_x, next_y);
        }
    }
}

File: 백준/Silver/21736. 헌내기는 친구가 필요해/헌내기는 친구가 필요해.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

n, m = map(int, input().split())
field = [input() for _ in range(n)]
visited = [[False]*m for _ in range(n)]

dx = [0, 0, -1, 1]
dy = [1, -1, 0, 0]
count = 0

def is_valid(x, y):
    return 0 <= x < n and 0 <= y < m and not visited[x][y]

def dfs(x, y):
    global count
    visited[x][y] = True
    if field[x][y] == 'P':
        count += 1
    for i in range(4):
        nx, ny = x + dx[i], y + dy[i]
        if is_valid(nx, ny) and field[nx][ny] != 'X':
            dfs(nx, ny)

# 시작 위치 찾기
for i in range(n):
    for j in range(m):
        if field[i][j] == 'I':
            si, sj = i, j

dfs(si, sj)
print(count if count > 0 else "TT")